hbase_sql/src/main/java/com/splicemachine/lifecycle/RegionServerLifecycle.java

```diff
@@ -144,6 +146,8 @@ private boolean causeIsPleaseHold(Throwable e) {
                 if (PleaseHoldException.class.getName().equals(re.getClassName()))
                     return true;
             }
+            return (e.getCause() instanceof IOException && e.getCause().getCause() instanceof CallTimeoutException) ||
+                    e.getCause() instanceof RemoteWithExtrasException;
```

Comment:
The last condition seems to be very general, if the cause is RemoteWithExtrasException we'd assume the Master is not initialized but we don't inspect what caused the RemoteWithExtrasException. What case is that condition handling?

Reply:
This can handle some of the conditions, the first is in the customer's log, which doesn't have more causes:

```
Caused by: org.apache.hadoop.hbase.ipc.RemoteWithExtrasException(java.io.IOException): java.io.IOException: Table Namespace Manager not fully initialized, try again later
        at org.apache.hadoop.hbase.master.HMaster.checkNamespaceManagerReady(HMaster.java:2753)
        at org.apache.hadoop.hbase.master.HMaster.ensureNamespaceExists(HMaster.java:2976)
        at org.apache.hadoop.hbase.master.HMaster.createTable(HMaster.java:1746)
```

The second is what I found during test, which has another cause:

```
Caused by: org.apache.hadoop.hbase.ipc.RemoteWithExtrasException(java.io.IOException): java.io.IOException: Unexpected exception
        at com.splicemachine.si.data.hbase.coprocessor.CoprocessorUtils.getIOException(CoprocessorUtils.java:30)
        at com.splicemachine.hbase.SpliceMasterObserver.preCreateTable(SpliceMasterObserver.java:142)
        at org.apache.hadoop.hbase.master.MasterCoprocessorHost$11.call(MasterCoprocessorHost.java:221)
        at org.apache.hadoop.hbase.master.MasterCoprocessorHost.execOperation(MasterCoprocessorHost.java:1151)
        at org.apache.hadoop.hbase.master.MasterCoprocessorHost.preCreateTable(MasterCoprocessorHost.java:217)
        at org.apache.hadoop.hbase.master.HMaster.createTable(HMaster.java:1578)
        at org.apache.hadoop.hbase.master.MasterRpcServices.createTable(MasterRpcServices.java:469)
        at org.apache.hadoop.hbase.protobuf.generated.MasterProtos$MasterService$2.callBlockingMethod(MasterProtos.java:55682)
        at org.apache.hadoop.hbase.ipc.RpcServer.call(RpcServer.java:2183)
        at org.apache.hadoop.hbase.ipc.CallRunner.run(CallRunner.java:112)
        at org.apache.hadoop.hbase.ipc.RpcExecutor$Handler.run(RpcExecutor.java:183)
        at org.apache.hadoop.hbase.ipc.RpcExecutor$Handler.run(RpcExecutor.java:163)
Caused by: java.lang.IllegalStateException: Startup failed
        at com.splicemachine.hbase.SpliceMasterObserver.preCreateTable(SpliceMasterObserver.java:138)
        ... 10 more

        at org.apache.hadoop.hbase.ipc.RpcClientImpl.call(RpcClientImpl.java:1269)
        at org.apache.hadoop.hbase.ipc.AbstractRpcClient.callBlockingMethod(AbstractRpcClient.java:227)
        at org.apache.hadoop.hbase.ipc.AbstractRpcClient$BlockingRpcChannelImplementation.callBlockingMethod(AbstractRpcClient.java:336)
        at 
```

I don't know if these two are the only errors which will occurs during the startup, so I handle `RemoteWithExtrasException`.

Reply:
The first case doesn't have another 'cause', but you can inspect the exception for the message "Table Namespace Manager not fully initialized" for instance.

The second error shouldn't be handled by waiting. The second error means that the Master initialization failed for some reason, so it will never become available. Waiting won't help here. That's why handling RemoteWithExtrasException just by waiting is wrong, in some cases we have to fail.